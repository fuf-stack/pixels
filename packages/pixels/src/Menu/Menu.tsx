import type { TVClassName, TVProps } from '@fuf-stack/pixel-utils';
import type { ButtonProps } from '@heroui/button';
import type {
  DropdownItemProps as HeroDropdownItemProps,
  DropdownProps as HeroDropdownProps,
  DropdownSectionProps as HeroDropdownSectionProps,
} from '@heroui/dropdown';
import type { Key, ReactNode } from 'react';

import { FaEllipsisVertical } from 'react-icons/fa6';

import { Button } from '@heroui/button';
import {
  Dropdown as HeroDropdown,
  DropdownItem as HeroDropdownItem,
  DropdownMenu as HeroDropdownMenu,
  DropdownSection as HeroDropdownSection,
  DropdownTrigger as HeroDropdownTrigger,
} from '@heroui/dropdown';

import { cn, tv, variantsToClassNames } from '@fuf-stack/pixel-utils';

/**
 * Menu item type
 */
export interface MenuItem {
  /** unique identifier */
  key: string;
  /** CSS class name */
  className?: string;
  /** additional description shown under the label */
  description?: string;
  /** disables the menu item */
  disabled?: boolean;
  /** menu item icon  */
  icon?: ReactNode;
  /** menu item name */
  label: ReactNode;
  /** click event handler */
  onClick?: HeroDropdownItemProps['onPress'];
  /** e2e test identifier */
  testId?: string;
}

/**
 * Menu section type
 */
export interface MenuSection {
  /** unique identifier */
  key: string;
  /** section label */
  label: ReactNode;
  /** section items (array of MenuItem) */
  items: MenuItem[];
}

// menu styling variants
export const menuVariants = tv({
  slots: {
    item: '',
    trigger: '',
  },
});

type VariantProps = TVProps<typeof menuVariants>;
type ClassName = TVClassName<typeof menuVariants>;

export interface MenuProps extends VariantProps {
  /** sets HTML aria-label attribute */
  ariaLabel?: string;
  /** child components */
  children?: ReactNode;
  /** CSS class name */
  className?: ClassName;
  /** menu item structure */
  items: (MenuSection | MenuItem)[];
  /** disable menu trigger */
  isDisabled?: boolean;
  /** placement of the menu */
  placement?: HeroDropdownProps['placement'];
  /** called if item is selected */
  onAction?: (key: Key) => void;
  /** HTML data-testid attribute used in e2e tests */
  testId?: string;
  /** When defined a Button will be rendered as trigger (with provided props) instead of unstyled html button */
  triggerButtonProps?: Pick<
    ButtonProps,
    | 'aria-label'
    | 'className'
    | 'color'
    | 'disableAnimation'
    | 'disabled'
    | 'size'
    | 'variant'
  >;
}

/** returns String[] of disabled items/keys */
const getDisabledKeys = (items: (MenuSection | MenuItem)[]) => {
  return (
    items
      // @ts-expect-error typing issue with MenuSection | MenuItem
      .map((item) => (typeof item?.items === 'undefined' ? item : item.items))
      .flat<MenuItem[]>()
      .filter((item) => {
        return Object.hasOwn(item, 'disabled') && item.disabled === true;
      })
      .map((item) => item.key)
  );
};

const renderMenuItem = (item: MenuItem, itemClassName?: string) => (
  <HeroDropdownItem
    className={cn(itemClassName, item.className)}
    data-testid={item.testId || item.key}
    description={item.description}
    key={item.key}
    onPress={item.onClick}
    startContent={item.icon}
  >
    {item.label}
  </HeroDropdownItem>
);

/**
 * Dropdown menu component based on [HeroUI Dropdown](https://www.heroui.com//docs/components/dropdown)
 */
const Menu = ({
  ariaLabel = undefined,
  children = null,
  className: _className = undefined,
  isDisabled = false,
  items,
  onAction = undefined,
  placement = undefined,
  testId = undefined,
  triggerButtonProps = undefined,
}: MenuProps) => {
  // className from slots
  const variants = menuVariants();
  const className = variantsToClassNames(variants, _className, 'trigger');

  // determine trigger button variant
  let triggerButton = <button type="button">{children}</button>;
  if (!children) {
    //  default to ellipsis icon when no children are provided
    triggerButton = (
      <Button
        className={cn('min-w-0', className.trigger)}
        size="sm"
        variant="flat"
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...triggerButtonProps}
      >
        <FaEllipsisVertical />
      </Button>
    );
  } else if (triggerButtonProps) {
    // use provided triggerButtonProps with hero button
    triggerButton = (
      // eslint-disable-next-line react/jsx-props-no-spreading
      <Button className={className.trigger} {...triggerButtonProps}>
        {children}
      </Button>
    );
  }

  return (
    <HeroDropdown
      aria-label={ariaLabel}
      isDisabled={isDisabled}
      placement={placement}
    >
      <HeroDropdownTrigger data-testid={testId}>
        {/* NOTE: type and aria properties are injected by HeroDropdownTrigger */}
        {triggerButton}
      </HeroDropdownTrigger>
      <HeroDropdownMenu
        items={items}
        disabledKeys={getDisabledKeys(items)}
        onAction={onAction}
      >
        {(item) => {
          if ('items' in item) {
            return (
              <HeroDropdownSection
                items={item.items}
                title={item.label as HeroDropdownSectionProps['title']}
                key={item.key}
              >
                {(sectionItem) => {
                  return renderMenuItem(sectionItem, className.item);
                }}
              </HeroDropdownSection>
            );
          }
          return renderMenuItem(item, className.item);
        }}
      </HeroDropdownMenu>
    </HeroDropdown>
  );
};

export default Menu;
