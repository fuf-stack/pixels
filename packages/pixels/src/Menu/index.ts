import Menu from './Menu';

export type { MenuProps } from './Menu';

export { Menu };

export default Menu;
