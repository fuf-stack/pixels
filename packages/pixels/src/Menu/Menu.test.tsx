import { describe, expect, test } from 'vitest';

import storySnapshots from '@repo/storybook-config/story-snapshots';
import { fireEvent, render } from '@testing-library/react';

import Menu from './Menu';
import * as stories from './Menu.stories';

describe('Story Snapshots', () => {
  storySnapshots(stories);
});

describe('Coverage', () => {
  test('renders Menu items correctly', () => {
    const { container, getByText } = render(
      <Menu items={[{ label: 'testItem', key: 'testitem' }]} />,
    );

    // Click the button
    fireEvent.click(container.firstChild);

    // Check if the error is shown
    const items = getByText('testItem');
    expect(items).toBeInTheDocument();
  });
  test('renders Menu items of item correctly', () => {
    const { container, getByText } = render(
      <Menu
        items={[
          { label: 'firstItem', key: 'firstitem' },
          {
            label: 'SubMenu',
            key: 'submenu',
            items: [{ label: 'subItem', key: 'subitem' }],
          },
        ]}
      />,
    );

    // Click the button
    fireEvent.click(container.firstChild);

    const firstItem = getByText('firstItem');
    expect(firstItem).toBeInTheDocument();

    const subMenu = getByText('SubMenu');
    expect(subMenu).toBeInTheDocument();

    const subItem = getByText('subItem');
    expect(subItem).toBeInTheDocument();
  });
});
