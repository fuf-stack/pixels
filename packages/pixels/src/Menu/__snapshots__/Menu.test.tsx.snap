// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Story Snapshots > AllPlacements 1`] = `
<div>
  <div
    class="grid grid-cols-3 gap-4"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«ri»"
      tabindex="0"
      type="button"
    >
      top
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«rl»"
      tabindex="0"
      type="button"
    >
      top-start
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«ro»"
      tabindex="0"
      type="button"
    >
      top-end
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«rr»"
      tabindex="0"
      type="button"
    >
      bottom
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«ru»"
      tabindex="0"
      type="button"
    >
      bottom-start
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r11»"
      tabindex="0"
      type="button"
    >
      bottom-end
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r14»"
      tabindex="0"
      type="button"
    >
      left
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r17»"
      tabindex="0"
      type="button"
    >
      left-start
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r1a»"
      tabindex="0"
      type="button"
    >
      left-end
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r1d»"
      tabindex="0"
      type="button"
    >
      right
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r1g»"
      tabindex="0"
      type="button"
    >
      right-start
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
      data-react-aria-pressable="true"
      data-slot="trigger"
      id="react-aria-«r1j»"
      tabindex="0"
      type="button"
    >
      right-end
    </button>
  </div>
</div>
`;

exports[`Story Snapshots > Default 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r0»"
    tabindex="0"
    type="button"
  >
    <span
      data-testid="icon-FaEllipsisVertical"
    >
      FaEllipsisVertical
    </span>
  </button>
</div>
`;

exports[`Story Snapshots > Disabled 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased opacity-disabled pointer-events-none min-w-0"
    data-disabled="true"
    data-react-aria-pressable="true"
    data-slot="trigger"
    disabled=""
    id="react-aria-«rf»"
    type="button"
  >
    <span
      data-testid="icon-FaEllipsisVertical"
    >
      FaEllipsisVertical
    </span>
  </button>
</div>
`;

exports[`Story Snapshots > MenuParty 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Open fun menu"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-secondary text-secondary data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r1v»"
    tabindex="0"
    type="button"
  >
    Open fun menu 🎈
  </button>
</div>
`;

exports[`Story Snapshots > SubMenu 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«rc»"
    tabindex="0"
    type="button"
  >
    <span
      data-testid="icon-FaEllipsisVertical"
    >
      FaEllipsisVertical
    </span>
  </button>
</div>
`;

exports[`Story Snapshots > WithAriaLabelAndTestIds 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0"
    data-react-aria-pressable="true"
    data-slot="trigger"
    data-testid="menu-trigger"
    id="react-aria-«r1p»"
    tabindex="0"
    type="button"
  >
    <span
      data-testid="icon-FaEllipsisVertical"
    >
      FaEllipsisVertical
    </span>
  </button>
</div>
`;

exports[`Story Snapshots > WithOnAction 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r1m»"
    tabindex="0"
    type="button"
  >
    <span
      data-testid="icon-FaEllipsisVertical"
    >
      FaEllipsisVertical
    </span>
  </button>
</div>
`;

exports[`Story Snapshots > WithSlotClassNames 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased bg-black text-white hover:bg-gray-900"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r1s»"
    tabindex="0"
    type="button"
  >
    Styled trigger
  </button>
</div>
`;

exports[`Story Snapshots > WithTrigger 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r6»"
    tabindex="0"
    type="button"
  >
    trigger
  </button>
</div>
`;

exports[`Story Snapshots > WithTriggerButtonProps 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Open actions menu"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] !transition-none data-[pressed=true]:scale-100 bg-transparent border-primary text-primary data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r9»"
    tabindex="0"
    type="button"
  >
    Open menu
  </button>
</div>
`;

exports[`Story Snapshots > WithoutTrigger 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-solid outline-transparent data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default/40 text-default-700 data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased min-w-0"
    data-react-aria-pressable="true"
    data-slot="trigger"
    id="react-aria-«r3»"
    tabindex="0"
    type="button"
  >
    <span
      data-testid="icon-FaEllipsisVertical"
    >
      FaEllipsisVertical
    </span>
  </button>
</div>
`;
